blog: crs migration part 5 (#505)
coreruleset/website: content/blog/2026-04-27-migrating-from-crs-3-to-crs-4-part-5-rule-changes.md

## New: HTTP/3 Support

CRS 4 adds rules that correctly handle HTTP/3 semantics. For most operators this is invisible — your WAF engine handles the protocol level, not CRS. But if you use a WAF engine that passes HTTP/3-specific headers or pseudo-headers to the rule set, CRS 4 handles them correctly where CRS 3 would have misbehaved or missed them.

## Changed: HTTP/0.9 Tolerance Dropped

HTTP/0.9 was already absent from the default `tx.allowed_http_versions` list in CRS 3.3, so the protocol enforcement rule in `REQUEST-920` has been blocking it by default in both versions. What changed in CRS 4 is narrower: response-splitting detection (rule `921110`) used to carve out an HTTP/0.9 pattern to avoid false positives on legacy clients, and that carve-out was removed. See the "drop HTTP/0.9 support" change in [CHANGES.md](https://github.com/coreruleset/coreruleset/blob/v4.0/dev/CHANGES.md) (PR #1966).

> For most operators this is invisible at migration time. If you explicitly allowed HTTP/0.9 in CRS 3 by adding it to `tx.allowed_http_versions`, the same configuration continues to work in CRS 4, but rule `921110` will no longer tolerate HTTP/0.9 in its regex. Monitoring agents that still speak HTTP/0.9 should be updated to HTTP/1.0 or later regardless.

Author: Need to update this now that we will backport https://github.com/coreruleset/coreruleset/pull/4621

- Rules that now fire where they did not in CRS 3 (new exclusions needed)
- Exclusions that are no longer needed (the underlying false positive was fixed)
- Score changes at your paranoia level due to PL redistribution

This step cannot be skipped. The combination of rule reorganization and PL redistribution means your anomaly score baseline will be different from CRS 3, even if you carefully map every exclusion.

## Rule File Organization

CRS 4 preserves the familiar `REQUEST-9xx-*.conf` and `RESPONSE-9xx-*.conf` naming convention. The numbering within each file and some file boundaries changed, but the overall organization is similar enough that you can navigate CRS 4 files if you know CRS 3.

A notable addition: CRS 4 ships with `REQUEST-900-EXCLUSION-RULES-BEFORE-CRS.conf.example` and `RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf.example` as officially supported places to add your custom exclusions. These files are the CRS-blessed approach to exclusion management. If your current exclusions are scattered across your WAF configuration, consider consolidating them into these files during the migration.

## What's Next

> [Part 6]({{< ref "blog/2026-05-04-migrating-from-crs-3-to-crs-4-part-6-tuning.md" >}}) covers the practical approach to false positive tuning during the migration — two strategies for carrying over your existing tuning, how to run a production migration safely, and common false positives that changed between versions.

Author: We'll keep this one for now.

slug: 'migrating-crs-3-to-4-part-5-rule-changes'
---

This is Part 5 of the [CRS 3.3 → 4.25 LTS migration series]({{< ref "blog/2026-03-30-migrating-from-crs-3-to-crs-4-part-1-overview.md" >}}). Part 4 covered anomaly scoring and reporting. This post covers the rule-level changes: what is new, what is gone, what moved, and how to audit your existing exclusions against the CRS 4 rule set.

{{< figure src="/images/2026/04/pexels-egorkomarov-8824026.jpg" caption="Hundreds of rules changed under the hood" attr="Egor Komarov on Pexels" attrlink="https://www.pexels.com" >}}

## The Scale of Change

There are hundreds of rule-level changes between CRS 3.3 and CRS 4.0. This is not a point release — it is the result of years of accumulated improvements, a public bug bounty programme, and deliberate architectural cleanup. Understanding the shape of this change helps you plan your tuning work.

The changes fall into four categories:

1. **New rules and new detection categories** — coverage that did not exist in CRS 3
2. **Modified rules** — rules that exist in both versions but with changed patterns, PL assignments, or IDs

> 3. **Removed rules** — rules that were dropped because they were superseded, redundant, or false-positive-prone

Reviewer: ```suggestion
3. **Removed rules** — rules that were dropped because they were superseded, redundant, or prone to false positives
```

4. **Reorganized rules** — rules that were split, merged, or renumbered

> The authoritative source for all of this is the [CHANGES.md](https://github.com/coreruleset/coreruleset/blob/main/CHANGES.md) in the CRS 4.0 release. This post walks through the most significant categories. For a full audit of your specific exclusions, you will need to cross-reference CHANGES.md with your existing configuration.

Reviewer: ```suggestion
The authoritative source for all of these changes is the [CHANGES.md](https://github.com/coreruleset/coreruleset/blob/main/CHANGES.md) in the CRS 4.0 release. This post walks through the most significant categories. For a full audit of your specific exclusions, you will need to cross-reference CHANGES.md with your existing configuration.
```

> ### A note on the CRS 4.25 LTS cadence

Reviewer: ```suggestion
### A Note on the CRS 4.25 LTS Cadence
```

## Changed: HTTP/0.9 Tolerance Dropped

HTTP/0.9 has been progressively removed from CRS over several releases. The protocol is obsolete per RFC 9110 and is absent from the default `tx.allowed_http_versions` list — but because HTTP/0.9 requests carry no protocol string at all, the version-list check in rule `920430` only catches them when the engine populates `REQUEST_PROTOCOL` with a default. Two other rules did the real work of blocking HTTP/0.9, and both have been tightened:

- **CRS 4.0** — response-splitting detection (rule `921110`) used to carve out an HTTP/0.9 pattern to avoid false positives on legacy clients, and that carve-out was removed. See the "drop HTTP/0.9 support" change in [CHANGES.md](https://github.com/coreruleset/coreruleset/blob/main/CHANGES.md) (PR #1966).

> - **CRS 4.25.1** (first quarterly LTS backport, scheduled) — request-line validation (rule `920100`) had a dedicated alternative that accepted `GET /path` without any protocol suffix, preserving HTTP/0.9 compatibility. That alternative was removed (PR [#4621](https://github.com/coreruleset/coreruleset/pull/4621)) on main and will be backported to the 4.25 LTS line in the first quarterly 4.25.1 release. After this change, `GET /\r\n\r\n` style requests trigger `920100` and are blocked.

Reviewer: ```suggestion
- **CRS 4.25.1** (first quarterly LTS backport, scheduled) — request-line validation (rule `920100`) had a dedicated alternative that accepted `GET /path` without any protocol suffix, preserving HTTP/0.9 compatibility. That alternative was removed (PR [#4621](https://github.com/coreruleset/coreruleset/pull/4621)) on main and will be backported to the 4.25 LTS line in the first quarterly 4.25.1 release. After this change, `GET /\r\n\r\n` style requests  will trigger `920100` and will be blocked.
```

- **Removed**: the rule no longer exists. Your exclusion is now a no-op. If you added it to fix a false positive, that false positive may have been fixed in CRS 4 — or the rule may have been renumbered. Test and remove the exclusion if it is no longer needed.
- **Renumbered**: the rule exists under a new ID. Update your exclusion to reference the new ID.
- **Split**: the rule was broken into multiple rules with new IDs. Update your exclusion to cover the relevant new IDs.
- **Moved to plugin**: the rule is now in a plugin, not in core. You need to install the plugin and the exclusion still applies — but now against the plugin's version of the rule.

### Step 3: Verify the Remaining Exclusions in Detection Mode

After updating your exclusions for renamed and removed rules, run CRS 4 in detection mode (anomaly scoring only, no blocking) for a representative period. This reveals:

- Rules that now fire where they did not in CRS 3 (new exclusions needed)
- Exclusions that are no longer needed (the underlying false positive was fixed)
- Score changes at your paranoia level due to PL redistribution

> This step cannot be skipped. The combination of rule reorganization and PL redistribution means your anomaly score baseline will be different from CRS 3, even if you carefully map every exclusion.

Reviewer: ```suggestion
This step must not be skipped. The combination of rule reorganization and PL redistribution means your anomaly score baseline will be different from CRS 3, even if you carefully map every exclusion.
```

> HTTP/0.9 has been progressively removed from CRS over several releases. The protocol is obsolete per RFC 9110 and is absent from the default `tx.allowed_http_versions` list — but because HTTP/0.9 requests carry no protocol string at all, the version-list check in rule `920430` only catches them when the engine populates `REQUEST_PROTOCOL` with a default. Two other rules did the real work of blocking HTTP/0.9, and both have been tightened:

Reviewer: Link to the RFC

Author: @copilot add link to the RFC here
